File: src/mlp_down.zig
const std = @import("std");

const addition = @import("simd/addition.zig");
const hadamard_product = @import("simd/hadamard_product.zig");
const matrix_vector_product = @import("simd/matrix_vector_product.zig");
const rms_norm = @import("simd/rms_norm.zig");
const swish = @import("simd/swish.zig");

const Self = @This();

down_weight: []f32,
input_vector: []f32,
output_vector: []f32,
residual_vector: []f32,

export fn create(input_size: usize, output_size: usize, num_layers: usize) ?*Self {
    const allocator = std.heap.page_allocator;
    const self = allocator.create(Self) catch return null;

    self.* = .{
        .down_weight = allocator.alloc(f32, num_layers * output_size * input_size) catch return null,
        .input_vector = allocator.alloc(f32, input_size) catch return null,
        .output_vector = allocator.alloc(f32, output_size) catch return null,
        .residual_vector = allocator.alloc(f32, output_size) catch return null,
    };

    return self;
}

export fn getDownWeight(self: *const Self) [*]f32 {
    return self.down_weight.ptr;
}

export fn getInputVector(self: *const Self) [*]f32 {
    return self.input_vector.ptr;
}

export fn getOutputVector(self: *const Self) [*]f32 {
    return self.output_vector.ptr;
}

export fn getResidualVector(self: *const Self) [*]f32 {
    return self.residual_vector.ptr;
}

export fn forward(self: *const Self, layer: usize) void {
    const weight_size = self.output_vector.len * self.input_vector.len;

    matrix_vector_product.compute(
        self.down_weight[layer * weight_size ..][0..weight_size],
        self.input_vector,
        self.output_vector,
    );

    addition.compute(self.output_vector, self.residual_vector, self.output_vector);
}

File: packages/decoder/src/mlp_down.ts
import { checkNotNull } from './check_not_null.js';

export interface MlpDownInit {
  readonly inputSize: number;
  readonly outputSize: number;
  readonly numLayers: number;
}

export class MlpDown {
  static wasmModule: WebAssembly.Module;

  static async create(init: MlpDownInit): Promise<MlpDown> {
    return new MlpDown(init, await WebAssembly.instantiate(this.wasmModule));
  }

  readonly downWeight: Uint8Array;
  readonly inputVector: Float32Array;
  readonly outputVector: Float32Array;
  readonly residualVector: Float32Array;

  private readonly self: number;

  constructor(
    { inputSize, outputSize, numLayers }: MlpDownInit,
    private readonly wasmInstance: WebAssembly.Instance,
  ) {
    const { memory, create, getDownWeight, getInputVector, getOutputVector, getResidualVector } =
      wasmInstance.exports as WasmExports;

    this.self = create(inputSize, outputSize, numLayers);

    this.downWeight = new Uint8Array(
      memory.buffer,
      checkNotNull(getDownWeight(this.self)),
      numLayers * outputSize * inputSize * 4,
    );

    this.inputVector = new Float32Array(
      memory.buffer,
      checkNotNull(getInputVector(this.self)),
      inputSize,
    );

    this.outputVector = new Float32Array(
      memory.buffer,
      checkNotNull(getOutputVector(this.self)),
      outputSize,
    );

    this.residualVector = new Float32Array(
      memory.buffer,
      checkNotNull(getResidualVector(this.self)),
      outputSize,
    );
  }

  forward({ layer }: Readonly<{ layer: number }>): void {
    (this.wasmInstance.exports as WasmExports).forward(this.self, layer);
  }
}

interface WasmExports extends WebAssembly.Exports {
  readonly memory: WebAssembly.Memory;

  create(inputSize: number, outputSize: number, numLayers: number): number;
  getDownWeight(self: number): number;
  getInputVector(self: number): number;
  getOutputVector(self: number): number;
  getResidualVector(self: number): number;
  forward(self: number, layer: number): void;
}
